### PR TITLE
[Merged by Bors] - feat: an irreducible real polynomial has degree ≤2

### PR DESCRIPTION
**Pull request by urkud**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by grunweg — APPROVED** [review]
Looks plausible to me. The proof is surprisingly readable!
(confidence low; I'm barely know this area of mathlib)

File: Mathlib/Data/IsROrC/Lemmas.lean [hunk 1]
```diff
@@ -87,3 +87,14 @@ theorem ofRealCLM_norm : ‖(ofRealCLM : ℝ →L[ℝ] K)‖ = 1 :=
 #align is_R_or_C.of_real_clm_norm IsROrC.ofRealCLM_norm
 
 end IsROrC
+
+namespace Polynomial
+
+open ComplexConjugate in
+lemma aeval_conj (p : ℝ[X]) (z : K) : aeval (conj z) p = conj (aeval z p) :=
+  aeval_algHom_apply (@IsROrC.conjAe K _ : K →ₐ[ℝ] K) z p
```
Comment (adomani) [inline Mathlib/Data/IsROrC/Lemmas.lean:95]:
```suggestion
  aeval_algHom_apply (IsROrC.conjAe (K := K)) z p
```
this is also an option.

File: Mathlib/Analysis/Complex/Polynomial.lean [hunk 2]
```diff
@@ -46,3 +48,44 @@ instance isAlgClosed : IsAlgClosed ℂ :=
...
+      map (algebraMap ℝ ℂ) (X ^ 2 - C (2 * z.re) * X + C (‖z‖ ^ 2))
+      _ = X ^ 2 - C (↑(2 * z.re) : ℂ) * X + C (‖z‖ ^ 2 : ℂ) := by simp
+      _ = (X - C (conj z)) * (X - C z) := by
+        rw [← add_conj, map_add, ← mul_conj', map_mul]
+        ring
+  rw [← map_dvd_map' (algebraMap ℝ ℂ), H]
+  apply IsCoprime.mul_dvd
+  · exact isCoprime_X_sub_C_of_isUnit_sub <| .mk0 _ <| sub_ne_zero.2 <| mt conj_eq_iff_im.1 hz
+  · simpa [dvd_iff_isRoot, aeval_conj]
+  · simpa [dvd_iff_isRoot]
```
Comment (adomani) [inline Mathlib/Analysis/Complex/Polynomial.lean:67]:
```suggestion
lemma Polynomial.mul_star_dvd_of_aeval_eq_zero_im_ne_zero (p : ℝ[X]) {z : ℂ} (h0 : aeval z p = 0)
    (hz : z.im ≠ 0) : (X - C ((starRingEnd ℂ) z)) * (X - C z) ∣ map (algebraMap ℝ ℂ) p := by
  apply IsCoprime.mul_dvd
  · exact isCoprime_X_sub_C_of_isUnit_sub <| .mk0 _ <| sub_ne_zero.2 <| mt conj_eq_iff_im.1 hz
  · simpa [dvd_iff_isRoot, aeval_conj]
  · simpa [dvd_iff_isRoot]

/-- If `z` is a non-real complex root of a real polynomial,
then `p` is divisible by a quadratic polynomial. -/
lemma Polynomial.quadratic_dvd_of_aeval_eq_zero_im_ne_zero (p : ℝ[X]) {z : ℂ} (h0 : aeval z p = 0)
    (hz : z.im ≠ 0) : X ^ 2 - C (2 * z.re) * X + C (‖z‖ ^ 2) ∣ p := by
  rw [← map_dvd_map' (algebraMap ℝ ℂ)]
  convert p.mul_star_dvd_of_aeval_eq_zero_im_ne_zero h0 hz
  calc
    map (algebraMap ℝ ℂ) (X ^ 2 - C (2 * z.re) * X + C (‖z‖ ^ 2))
    _ = X ^ 2 - C (↑(2 * z.re) : ℂ) * X + C (‖z‖ ^ 2 : ℂ) := by simp
    _ = (X - C (conj z)) * (X - C z) := by
      rw [← add_conj, map_add, ← mul_conj', map_mul]
      ring
```
I think that the lemma about divisibility by the product of the conjugates can be useful as a stand-alone.

**Review by adomani — COMMENTED** [review]
Very clean proof!

**Comment by adomani** [comment]
Thanks!

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**